source/ddmp/diff.d

```diff
@@ -289,24 +288,45 @@ if (isSomeString!Str) {
     }
 }
 
-/// The character offset should be aware of unicode, otherwise, the common prefix can end up
-/// splitting a single character's bytes.
 size_t commonPrefix(Str)(Str text1, Str text2)
-if (isSomeString!Str) {
-    auto n = min(text1.length, text2.length);
-    foreach (i; 0 .. n)
-        if (text1[i] != text2[i])
-            return i;
-    return n;
+{
+    size_t ret = 0;
+    while (text1.length && text2.length) {
+        size_t n = text1.utfStride;
```

Comment:
Yeah, these functions are fine so far. `halfMatch` is a bit more tricky, because it splits up the input string with a fixed ratio, but when looking at it for a second time, it quickly became clear that it can also be trivially made to work with no complexity changes due to the nature of the UTF encoding. The approach here would be to simple step back up to 3 (UTF-8) or 1 (UTF-16) code units and check whether that is the start of a valid UTF sequence and adjust the splitting index accordingly.

Unfortunately, after `halfMatch` was fixed, `bisect` turned up as the next source for issues and proved to be more difficult to fix, because it applies indices derived from stepping over one input string to the other input string. Apart from that it was possible to fix all other places that could produce invalid indices, without changing the algorithm, and a lot of tests, including the line count test pass like that.

But for that last bit I had to look up the [original paper](https://neil.fraser.name/writing/diff/myers.pdf) for the algorithm, and, without some more serious research, I would say that this can only be reasonably fixed by either going with UCS-4/UTF-32, or by maintaining an index for each input string, which maps from code point index to code unit index, which would actually require more memory than the UTF-32 approach.

It would certainly be possible to fix this in a different way and still have valid diffs as a result, but results could then differ from the original algorithm, so I don't think it is advisable to go that route.